libgnucash/engine/Account.cpp

```diff
@@ -4392,6 +4392,111 @@ xaccAccountSetReconcileLastDate (Account *acc, time64 last_date)
+    acct = QOF_INSTANCE(acc);
+
+    for (gint i = 0; i < numdates; i++)
+    {
+        std::string num_str = std::to_string (i);
+        qof_instance_slot_path_delete
+            (acct, {KEY_RECONCILE_INFO, num_str, "statement-date"});
+        qof_instance_slot_path_delete
+            (acct, {KEY_RECONCILE_INFO, num_str, "ending-balance"});
+    }
+    qof_instance_slot_path_delete (acct, {KEY_RECONCILE_INFO, "num-dates"});
+}
+
+gint
+xaccAccountGetReconcileNumDates (const Account *acc)
+{
+    gint64 num = 0;
+    GValue v = G_VALUE_INIT;
+    g_return_val_if_fail(GNC_IS_ACCOUNT(acc), FALSE);
+    qof_instance_get_path_kvp (QOF_INSTANCE(acc), &v,
+                               {KEY_RECONCILE_INFO, "num-dates"});
+    if (G_VALUE_HOLDS_INT64 (&v))
+        num = g_value_get_int64 (&v);
+
+    return num;
+}
+
+void
+xaccAccountSetReconcileNumDates (Account *acc, gint num)
+{
+    GValue v = G_VALUE_INIT;
+    g_return_if_fail(GNC_IS_ACCOUNT(acc));
+
+    g_value_init (&v, G_TYPE_INT64);
+    g_value_set_int64 (&v, num);
+    xaccAccountBeginEdit (acc);
+    qof_instance_set_path_kvp (QOF_INSTANCE (acc), &v,
+                               {KEY_RECONCILE_INFO, "num-dates"});
+    mark_account (acc);
+    xaccAccountCommitEdit (acc);
+}
+
+
+gboolean
+xaccAccountGetReconcileDateBalance (const Account *acc,
+                                    gint num,
+                                    time64 *statement_date,
+                                    gnc_numeric *ending_balance)
+{
+    std::string num_str = std::to_string (num);
+    time64 date = 0;
+    gnc_numeric bal;
+    GValue v1 = G_VALUE_INIT, v2 = G_VALUE_INIT;
+    g_return_val_if_fail(GNC_IS_ACCOUNT(acc), FALSE);
+    qof_instance_get_path_kvp (QOF_INSTANCE(acc), &v1,
+                               {KEY_RECONCILE_INFO, num_str, "statement-date"});
+    qof_instance_get_path_kvp (QOF_INSTANCE(acc), &v2,
+                               {KEY_RECONCILE_INFO, num_str, "ending-balance"});
+
+    if (!(G_VALUE_HOLDS_INT64 (&v1) && G_VALUE_HOLDS_INT64 (&v2)))
+        return FALSE;
+
+    date = *(time64*) g_value_get_int64 (&v1);
+    bal = *(gnc_numeric*) g_value_get_boxed (&v2);
+
+    if (!bal.denom)
+        return FALSE;
+
+    if (statement_date) *statement_date = date;
+    if (ending_balance) *ending_balance = bal;
+    return TRUE;
+}
+
+void
+xaccAccountSetReconcileDateBalance (Account *acc, gint num, time64 statement_date,
+                                    gnc_numeric ending_balance)
+{
+    GValue v1 = G_VALUE_INIT ,v2 = G_VALUE_INIT;
+    std::string num_str = std::to_string (num);
+    g_return_if_fail(GNC_IS_ACCOUNT(acc));
+
+    g_value_init (&v1, G_TYPE_INT64);
+    g_value_init (&v2, GNC_TYPE_NUMERIC);
+    g_value_set_int64 (&v1, statement_date);
+    g_value_set_boxed (&v2, &ending_balance);
+
+    xaccAccountBeginEdit (acc);
+    qof_instance_set_path_kvp (QOF_INSTANCE (acc), &v1,
+                               {KEY_RECONCILE_INFO, num_str, "statement-date"});
```

Comment:
This data architecture doesn't make sense to me. I'd think you'd want a KEY_RECONCILE_STATEMENTS frame with entries having key statement-date (in ISO date form, ie 20200317 or 2020-03-17) and value statement balance (a gnc-numeric). Keeping track of the numbers is going to be cumbersome.

Reply:
Interesting... I'd copypasted from `xaccAccountGetReconcileLastDate` which doesn't convert to ISO. I'll redo copying from `xaccTransSetDateInternal` instead.

Reply:
I think you're missing the point.  It's not about a date string vs. time64, it's about the structure of the KVP.